app/src/org/commcare/android/tasks/TemplatePrinterTask.java

```diff
@@ -0,0 +1,253 @@
+
+import org.commcare.android.util.TemplatePrinterUtils;
+
+
+/**
+ * Asynchronous task for populating an html document with data.
+ *
+ * @author Richard Lu
+ * @author amstone
+ */
+public class TemplatePrinterTask extends AsyncTask<Void, Void, Integer> {
+
+    /**
+     * The 4 result codes that can be sent back by this task
+     */
+    public static final int SUCCESS = 0;
+    public static final int IO_ERROR = 1;
+    public static final int VALIDATION_ERROR_MUSTACHE = 2;
+    public static final int VALIDATION_ERROR_CHEVRON = 3;
+
+    /**
+     * Used to track which type of error (2 or 3 above) was encountered in validateString
+     */
+    private static int validationErrorType;
+
+    /**
+     * Used to track the string in the template file where a validation error was encountered
+     */
+    private String problemString;
+
+    /**
+     * The template file for this print action
+     */
+    private final File inputFile;
+
+    /**
+     * The path where the populated template should be saved to
+     */
+    private final String outputPath;
+
+    /**
+     * The mapping from keywords to case property values to be used in populating the template
+     */
+    private final Bundle values;
+
+    private final PopulateListener listener;
+
+
+    public TemplatePrinterTask(File input, String outputPath, Bundle values,
+                               PopulateListener listener) {
+        this.inputFile = input;
+        this.outputPath = outputPath;
+        this.values = values;
+        this.listener = listener;
+    }
+
+    /**
+     * Attempts to perform population of the template file, and throws the appropriate exception
+     * if encountering an error.
+     */
+    @Override
+    protected Integer doInBackground(Void... params) {
+        try {
+            populateHtml(inputFile, values);
+            return SUCCESS;
+        } catch (IOException e) {
+            return IO_ERROR;
+        } catch (RuntimeException e) {
+            problemString = e.getMessage();
+            return validationErrorType;
+        }
+    }
+
+    /**
+     * Receives the return value from doInBackground and proceeds accordingly
+     */
+    @Override
+    protected void onPostExecute(Integer result) {
+        listener.onFinished(result, problemString);
+    }
+
+    /**
+     * Populates an html print template based on the given set of key-value pairings
+     * and save the newly-populated template to a temp location
+     *
+     * @param input the html print template
+     * @param values the mapping of keywords to case property values
+     * @throws IOException
+     */
+    private void populateHtml(File input, Bundle values) throws IOException {
+        // Read from input file
+        // throws IOException
+        String fileText = TemplatePrinterUtils.docToString(input).toLowerCase();
+
+        // Check if <body></body> section of html string is properly formed
+        // throws RuntimeException
```

Comment:
Comments that describe what the code does are dangerous; what happens when the code changes? In this case it is easy to see what throws the exceptions because the methods used will have 'throws IOException' annotations on their declarations